silkworm/db/snapshots/bittorrent/web_session.cpp

```diff
@@ -0,0 +1,128 @@
+const std::map<std::string_view, std::string_view> kCloudflareHeaders{
+    {"lsjdjwcush6jbnjj3jnjscoscisoc5s", "I%OSJDNFKE783DDHHJD873EFSIVNI7384R78SSJBJBCCJBC32JABBJCBJK45"},
+};
+
+//! The timeout for HTTP asynchronous operations
+constexpr std::chrono::seconds kHttpTimeoutSecs{30};
+
+WebSession::WebSession(std::optional<std::string> server_certificate)
+    : server_certificate_(std::move(server_certificate)) {}
+
+Task<WebSession::StringResponse> WebSession::https_get(const urls::url& web_url, std::string_view target_file) const {
+    // The SSL context which holds root certificate used for verification ()
+    ssl::context ssl_ctx{ssl::context::tlsv13_client};
+    load_root_certificates(ssl_ctx, server_certificate_);
+    ssl_ctx.set_verify_mode(ssl::verify_peer);  // Ask to verify the remote server certificate
+
+    // These objects perform our I/O
+    net::ip::tcp::resolver resolver{co_await net::this_coro::executor};
+    beast::ssl_stream<beast::tcp_stream> ssl_stream{co_await net::this_coro::executor, ssl_ctx};
+    beast::tcp_stream& tcp_stream = beast::get_lowest_layer(ssl_stream);
+
+    // Set SNI Hostname (many hosts need this to handshake successfully)
+    const std::string host{web_url.host()};
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wold-style-cast"
+    if (!SSL_set_tlsext_host_name(ssl_stream.native_handle(), host.c_str())) {
+        beast::error_code ec{static_cast<int>(::ERR_get_error()), net::error::get_ssl_category()};
+        throw beast::system_error{ec, "error setting SNI hostname"};
+    }
+#pragma GCC diagnostic pop
+
+    const std::string port{web_url.has_port() ? web_url.port() : "443"};
+
+    // Look up the domain name
+    const auto resolve_results = co_await resolver.async_resolve(host, port, net::use_awaitable);
+
+    // Make the connection on the IP address we get from a lookup
+    tcp_stream.expires_after(kHttpTimeoutSecs);
+    co_await tcp_stream.async_connect(resolve_results, net::use_awaitable);
+
+    // Perform the SSL handshake
+    tcp_stream.expires_after(kHttpTimeoutSecs);
+    co_await ssl_stream.async_handshake(ssl::stream_base::client, net::use_awaitable);
+
+    // Setup the HTTP GET request message
+    http::request<http::empty_body> req{http::verb::get, target_file, kHttpVersion};
+    req.set(http::field::host, host);
+    req.set(http::field::user_agent, BOOST_BEAST_VERSION_STRING);
+    include_cloudflare_headers(req);
+    SILK_TRACE << "WebSeedClient::http_session HTTP request: " << req;
+
+    // Send the HTTP request to the remote host
+    tcp_stream.expires_after(kHttpTimeoutSecs);
+    const auto written_bytes = co_await http::async_write(ssl_stream, req, net::use_awaitable);
+    SILK_TRACE << "WebSeedClient::http_session HTTP request written_bytes: " << written_bytes;
+
+    // This buffer is used for reading
+    beast::flat_buffer data;
+
+    // Declare a container to hold the response
+    http::response<http::string_body> rsp;
```

Comment:
what's the relation between these? how come the data in flat_buffer survives after this returns?

Reply:
rsp -> response

Reply:
> what's the relation between these? how come the data in flat_buffer survives after this returns?

https://www.boost.org/doc/libs/1_86_0/libs/beast/doc/html/beast/ref/boost__beast__http__async_read/overload2.html